webapp-django/crashstats/crashstats/static/crashstats/js/socorro/correlation.js

```diff
@@ -97,4 +97,169 @@ $(document).ready(function () {
+            let value = view.getUint32(i);
+            // toString(16) will give the hex representation of the number without padding.
+            let stringValue = value.toString(16);
+            // We use concatenation and slice for padding.
+            let padding = '00000000';
+            let paddedValue = (padding + stringValue).slice(-padding.length);
+            hexCodes.push(paddedValue);
+        }
+
+        // Join all the hex strings into one
+        return hexCodes.join('');
+    }
+
+    function sha1(str) {
+        return crypto.subtle.digest('SHA-1', new TextEncoder('utf-8').encode(str))
+        .then(hash => hex(hash));
+    }
+
+    let correlationData = {};
+
+    function getDataURL(product) {
+        if (product === 'Firefox') {
+            return 'https://analysis-output.telemetry.mozilla.org/top-signatures-correlations/data/';
+        } else if (product === 'FennecAndroid') {
+            return 'https://analysis-output.telemetry.mozilla.org/top-fennec-signatures-correlations/data/';
+        } else {
+            throw new Error('Unknown product: ' + product);
+        }
+    }
+
+    function loadChannelsData(product) {
+        if (correlationData[product]) {
+            return Promise.resolve();
+        }
+
+        return fetch(getDataURL(product) + 'all.json.gz')
+        .then(response => response.json())
+        .then(totals => {
+            correlationData[product] = {
+                'date': totals['date'],
+            };
+
+            for (let ch of ['release', 'beta', 'aurora', 'nightly']) {
+                correlationData[product][ch] = {
+                    'total': totals[ch],
+                    'signatures': {},
+                }
+            }
+        });
+    }
+
+    function loadCorrelationData(signature, channel, product) {
+        return loadChannelsData(product)
+        .then(() => {
+            if (signature in correlationData[product][channel]['signatures']) {
+                return;
+            }
+
+            return sha1(signature)
+            .then(sha1signature => {
+              return sha1signature;
+            })
+            .then(sha1signature => fetch(getDataURL(product) + channel + '/' + sha1signature + '.json.gz'))
+            .then(response => response.json())
+            .then(data => {
+                correlationData[product][channel]['signatures'][signature] = data;
+            });
+        })
+        .catch(() => {})
+        .then(() => correlationData);
+    }
+
+    function itemToLabel(item) {
+        return Object.getOwnPropertyNames(item)
+        .map(key => key + ' = ' + item[key])
+        .join(' ∧ ');
+    }
+
+    function toPercentage(num) {
+        let result = (num * 100).toFixed(2);
+
+        if (result == '100.00') {
+          return '100.0';
+        }
+
+        if (result.substring(0, result.indexOf('.')).length == 1) {
```

Comment:
``result`` being a number, that should raise a ``TypeError`` because (number).substring is not a function.